### PR TITLE
Separate lifetime for missing records for `ExpiringCache`

### LINKED ISSUE #539: Enhance expiring cache with per-record lifetime
The current expiring cache implementation has a lifetime value that is set in the constructor (https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/util/container/cache/ExpiringCache.d#L84) making it impossible to store different records for different lifetimes in the same cache. To do this currently would require multiple caches which complicates application logic.

It would be nice if different lifetimes for records in the same cache were possible so, for example, records expected to change frequently have a short lifetime and more stable records have a longer lifetime.

### PR DESCRIPTION
**Pull request by mihails-strasuns**
Fixes #539

### CODE REVIEW

File: relnotes/expiring.feature.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
+### Separate lifetime for empty records of ExpiringCache
+
+`ocean.util.container.cache.ExpiringCache`
+
+Now supports extra optional constructor argument - lifetime used for empty
+cached records. If not specified, regular lifetime value will be used, same as
+before.
+
+By empty record `ExpiringCache` considers one with internaly array length of 0.
```
Comment (gavin-norman-sociomantic) [inline relnotes/expiring.feature.md]: Wording:
"
`ExpiringCache` considers a cached record with array length 0 as being "empty".
"

File: src/ocean/util/container/cache/ExpiringCache.d [hunk 2]
```diff
@@ -54,13 +54,22 @@ class ExpiringCache ( size_t ValueSize = 0 ) : Cache!(ValueSize, true),
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/cache/ExpiringCache.d:58]: Suggest adding this to the module doc:
"
The cache can be configured to apply different lifetimes to empty and non-empty values (where a cached record with array length 0 is considered "empty"). See constructor arguments.
"

File: src/ocean/util/container/cache/ExpiringCache.d [hunk 3]
```diff
@@ -54,13 +54,22 @@ class ExpiringCache ( size_t ValueSize = 0 ) : Cache!(ValueSize, true),
...
+        Life time for non-empty items in seconds; may be changed at any time.
         This value must be at least 1.
 
     ***************************************************************************/
 
     public time_t lifetime;
 
+    /***************************************************************************
+
+        Life time for empty items in seconds; may be changed at any time.
```
Comment (gavin-norman-sociomantic) [inline src/ocean/util/container/cache/ExpiringCache.d:70]: Suggest defining what empty means here too.

**Review by gavin-norman-sociomantic — COMMENTED** [review]
LGTM. Couple of doc suggestions.

**Comment by mihails-strasuns** [comment]
Updated:
- fixed wording
- workaround to fix `-release` build

File: src/ocean/util/container/cache/ExpiringCache.d [hunk 4]
```diff
@@ -304,9 +323,6 @@ class ExpiringCache ( size_t ValueSize = 0 ) : Cache!(ValueSize, true),
     }
     body
     {
-        verify (this.lifetime > 0,
```
Comment (ben-palmer-sociomantic) [inline src/ocean/util/container/cache/ExpiringCache.d:307]: Removal of this check could be in its own commit.

File: src/ocean/util/container/cache/model/Value.d [hunk 5]
```diff
@@ -130,6 +130,18 @@ template Value ( size_t ValueSize )
...
             }
+
+            /*******************************************************************
+
+                Returns:
+                    length of underlying array
+
+            *******************************************************************/
+
+            public size_t length ( )
```
Comment (ben-palmer-sociomantic) [inline src/ocean/util/container/cache/model/Value.d:141]: Another option would be to have `empty` instead of `length` IIUC, of less general use though.

**Review by ben-palmer-sociomantic — COMMENTED** [review]
Just a couple of comments, the changes LGTM.

File: src/ocean/util/container/cache/ExpiringCache.d (hunk 4 above)
Comment (mihails-strasuns) [reply src/ocean/util/container/cache/ExpiringCache.d:307]: It is not removed, it is moved several lines below to check both lifetimes depending on item state.

Comment (ben-palmer-sociomantic) [reply src/ocean/util/container/cache/ExpiringCache.d:307]: Oh sorry, I thought you removed it because it was already checked in the constructor. Ignore my comment then.

File: src/ocean/util/container/cache/model/Value.d (hunk 5 above)
Comment (mihails-strasuns) [reply src/ocean/util/container/cache/model/Value.d:141]: Yeah I think `length` is better because it is a strict superset of functionality :)
